results/pkg/api/server/server.go

```diff
@@ -155,54 +163,186 @@ func (s *Server) ListResultsResult(ctx context.Context, req *pb.ListResultsReque
-	if req.GetFilter() == "" {
-		return &pb.ListResultsResponse{Results: results}, nil
+// Check if the result can be reserved.
+func matchCelFilter(r *pb.Result, prg cel.Program) (bool, error) {
+	if prg == nil {
+		return true, nil
+	}
+	for _, e := range r.Executions {
+		var (
+			taskrun     pb.TaskRun
+			pipelinerun pb.PipelineRun
+		)
+		if t := e.GetTaskRun(); t != nil {
+			taskrun = *(t)
+		}
+		if p := e.GetPipelineRun(); p != nil {
+			pipelinerun = *(p)
+		}
+		out, _, err := prg.Eval(map[string]interface{}{
+			"taskrun":     taskrun,
+			"pipelinerun": pipelinerun,
+		})
+		if err != nil && !strings.Contains(err.Error(), "no such attribute") && !strings.Contains(err.Error(), "undeclared reference to") {
+			log.Printf("failed to evaluate the expression: %v", err)
+			return false, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		}
+		if out.Value() == true {
+			return true, nil
+		}
 	}
+	return false, nil
+}
+
+// Encodes a result to PageToken
+func (s Server) encodeResult(primaryKey uint64) string {
+	var result []byte
+	if err := s.db.QueryRow("SELECT data from results where id=?", primaryKey).Scan(&result); err != nil {
+		return ""
+	}
+	encodedResult := make([]byte, base64.RawURLEncoding.EncodedLen(len(result)))
+	base64.RawURLEncoding.Encode(encodedResult, result)
+	return base64.RawURLEncoding.EncodeToString(encodedResult)
+}
 
-	// filter from all results
-	prg, err := s.env.Program(ast)
+func (s Server) decodeTokenToPrimaryKey(token string) (uint64, error) {
+	encodedToken, err := base64.RawURLEncoding.DecodeString(token)
+	if err != nil {
+		return 0, err
+	}
+	result := make([]byte, base64.RawURLEncoding.DecodedLen(len(encodedToken)))
+	_, err = base64.RawURLEncoding.Decode(result, encodedToken)
 	if err != nil {
-		log.Printf("program construction error: %s", err)
-		return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter checking step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
-	}
-	var resp []*pb.Result
-	for _, r := range results {
-		for _, e := range r.Executions {
-			out, _, err := prg.Eval(map[string]interface{}{
-				"taskrun": e.GetTaskRun(),
-			})
-			if err != nil {
-				log.Printf("failed to evaluate the expression: %v", err)
-				return nil, status.Errorf(codes.InvalidArgument, "Error occurred during filter evaluation step, no Results found for the query string due to invalid field, invalid function to evaluate filter or missing double quotes around field value, please try to enter a query with correct type again: %v", err)
+		return 0, err
+	}
+	p := &pb.Result{}
```

Comment:
We probably shouldn't use the full result proto as the pagination token to encode 1 value. Let's create a new proto message (in it's own internal package) that only contains the information we need.

Reply:
Oh, that's great!